### PR TITLE
Support for stm32

### PR DESCRIPTION
**Pull request by haydenth**
Hello @neilh10 - thanks for making this library. We made a few changes to the library to support stm32 along with a handful of bugfixes. You may want to take a look when you can. Thank you!

### CODE REVIEW

**Comment by neilh10** [comment]
THanks for the submission. 
Personally I have given up on the chip,  as I seem to remember some chips hung and didn't  recover- so perhaps your timeouts will solve that. I thought I had  enabled them at a higher level. It will take me a few days to set up a test system to try it.
See https://github.com/neilh10/STC3100arduino/issues/1 - if you have any data on why its hanging perhaps you could add to it. 

The V measurement is good, however the used mAHr in determining battery capacity didn't seem to really work out.
A field deployment with the initial version has worked  - search STC3100
https://monitormywatershed.org/sites/TUCA_PO03/

**Comment by haydenth** [comment]
Bringing in @fronders into the conversation here - I don't think we've run into any timeouts, have you noticed any?

**Comment by fronders** [comment]
These timeout adaptations are only valid for STM32 arch, since I2C transactions there use interrupts, and timeouts are managed automatically by Wire lib (the Wire lib doesn't have `getWireTimeoutFlag()` method).
So these changes only utilize checking for I2C errors on STM32 arch, which include timeouts by design.

Overall during testing haven't noticed any timeouts during usage of STC3100

**Comment by haydenth** [comment]
Well @neilh10 If you've abandoned this project, do you mind if I take it over?

File: src/STC3100dd.cpp
Comment (neilh10) [inline src/STC3100dd.cpp:1]: Suggest under #include "STC3100dd.h"

// For ARDUINO_ARCH_STM32 need to use specific API
// This uses an arm cc compiler, and excludes the ARDUINO WIRE timeouts
#if !defined(ARDUINO_ARCH_STM32)
//The default is to use the base ARDUINO WIRE TIMEOUT
#define ARDUINO_WIRE_TIMEOUT
#endif //ARDUINO_ARCH_STM32

#if defined(WIRE_HAS_TIMEOUT)
changed to 
#if defined(ARDUINO_WIRE_TIMEOUT)

**Comment by neilh10** [comment]
@haydenth - glad you want to take it over. I'm still using it for  test instrumentation - so not for the next year or so, but happy to make sure it still works for the base Arduino.  (I'm trying to move to ARM SAMD51 for the main processor - so have seen there are a few cc issues from the avr cc)
On your changes, I would want to see no changes for the base environment for the AVR compiler.
For your ARDUINO_ARCH_STM32 compiler I would propose at the beginning of 
STC3100dd.cpp

after #include "STC3100dd.h"
// For ARDUINO_ARCH_STM32 need to use specific API
// This uses an arm cc compiler, and excludes the ARDUINO WIRE timeouts
#if !defined(ARDUINO_ARCH_STM32)
//The default is to use the base ARDUINO WIRE TIMEOUT
#define ARDUINO_WIRE_TIMEOUT
#endif //ARDUINO_ARCH_STM32

**Comment by fronders** [comment]
@neilh10 the thing is, not all arches have `Wire.getWireTimeoutFlag()` in the Wire implementation and `WIRE_HAS_TIMEOUT` is specifically there to determine whether it is present or not, see the [bottom of the reference page](https://www.arduino.cc/reference/en/language/functions/communication/wire/getwiretimeoutflag/)

AVR environment provides that definition so this adds robustness to implementation as otherwise it would break on compilation on other arches. STM32 arch doesn't have one (since it doesn't have `Wire.getWireTimeoutFlag()`) so it enables STM32 timeout code if wire timeout is not available AND if the arch is STM32 (this also covers the case when potentially STM32 Wire implementation adds `Wire.getWireTimeoutFlag()` which I really doubt) 

I don't see a point introducing another define `ARDUINO_WIRE_TIMEOUT`

**Comment by neilh10** [comment]
@fronders  -thanks for the references. I get it now - the  WIRE_HAS_TIMEOUT is automatically defined, I hadn't realized that I thought it was need to be defined in the compiler to enable it.

**Comment by neilh10** [comment]
FYI - I don't get your changes for reading the serial number and that doesn't work for my board 
From the manual , Device ID registers 24 to 31 - the serial number is effectively Binary in REg ID1-6 ~ or 7 * 8bit numbers, 14 * 4bit hexadecimal numbers. 
REG_ID0 24 R Part type ID = 10h
REG_ID1 25 R Unique part ID, bits 0-7
REG_ID2 26 R Unique part ID, bits 8-15
REG_ID3 27 R Unique part ID, bits 16-23
REG_ID4 28 R Unique part ID, bits 24-31
REG_ID5 29 R Unique part ID, bits 32-39
REG_ID6 30 R Unique part ID, bits 40-47
REG_ID7 31 R Device ID CRC

My algorithm is slightly wrong. You're algorithm doesn't give me any useful information - am I missing something?
I'm looking at an updated section ..

        sn.reserve(2*STC3100_ID_LEN+2); //Number of Ascii characcters for 8*bytes
        #if !defined(ARDUINO_ARCH_STM32)
        for (int snlp=1;snlp<(STC3100_ID_LEN-1);snlp++) {
                //1st byte is ID (STC3100=0x10) last byte is CRC 
                uint8_t temp_num=serial_number[snlp];
                sn +=String((temp_num>>4) & 0x0f,HEX);
                sn +=String((temp_num & 0x0f),HEX);
        }
        #else
        for (int snlp = 1; snlp < STC3100_ID_LEN - 1; snlp++) {
            if (serial_number[snlp] < 0x10) {
                sn += '0';
            }
            sn += String(serial_number[snlp], HEX);
        }
        #endif

**Comment by neilh10** [comment]
For me it turns out that WIRE_HAS_TIMEOUT is not defined.
@fronders appreciate your insights - transferring this to a discussion
https://github.com/neilh10/STC3100arduino/discussions/4

**Comment by neilh10** [comment]
thanks explanation, I hadn't understood that first string test adding '0'

**Comment by haydenth** [comment]
Thank you guys!!

**Pull request merged**